monai/apps/pathology/transforms/post/array.py

```diff
@@ -0,0 +1,320 @@
+            )
+
+        distance_map = (1.0 - foreground_prob_map) * mask
+
+        if callable(self.smooth_fn):
+            distance_map = self.smooth_fn(distance_map)
+
+        return convert_to_dst_type(-distance_map, mask, dtype=self.dtype)[0]
+
+
+class GenerateMarkers(Transform):
+    """
+    Generate markers to be used in `watershed`. The watershed algorithm treats pixels values as a local topography
+    (elevation). The algorithm floods basins from the markers until basins attributed to different markers meet on
+    watershed lines. Generally, markers are chosen as local minima of the image, from which basins are flooded.
+    Here is the implementation from HoVerNet papar.
+    For more details refer to papers: https://arxiv.org/abs/1812.06499.
+
+    Args:
+        threshold: threshold the float values of foreground probability map to int 0 or 1 with specified theashold.
+            It turns uncertain area to 1 and other area to 0. Defaults to 0.4.
+        radius: the radius of the disk-shaped footprint used in `opening`. Defaults to 2.
+        min_size: objects smaller than this size are removed if `remove_small_objects` is True. Defaults to 10.
+        remove_small_objects: whether need to remove some objects in the marker. Defaults to True.
+        postprocess_fn: execute additional post transformation on marker. Defaults to None.
+        dtype: target data content type to convert, default is np.uint8.
+
+    """
+
+    backend = [TransformBackends.NUMPY]
+
+    def __init__(
+        self,
+        threshold: float = 0.4,
+        radius: int = 2,
+        min_size: int = 10,
+        remove_small_objects: bool = True,
+        postprocess_fn: Optional[Callable] = None,
+        dtype: DtypeLike = np.uint8,
+    ) -> None:
+        self.threshold = threshold
+        self.radius = radius
+        self.postprocess_fn = postprocess_fn
+        self.dtype = dtype
+
+        if remove_small_objects:
+            self.remove_small_objects = RemoveSmallObjects(min_size=min_size)
+
+    def __call__(self, mask: NdarrayOrTensor, foreground_prob_map: NdarrayOrTensor) -> NdarrayOrTensor:  # type: ignore
+        """
+        Args:
+            mask: binarized segmentation result. Shape must be [1, H, W].
+            foreground_prob_map: foreground probability map. Shape must be [1, H, W].
+        """
+        if mask.shape[0] != 1 or mask.ndim != 3:
+            raise ValueError(f"Input mask should be with size of [1, H, W], but got {mask.shape}")
+        if foreground_prob_map.shape[0] != 1 or foreground_prob_map.ndim != 3:
+            raise ValueError(
+                f"Input foreground_prob_map should be with size of [1, H, W], but got {foreground_prob_map.shape}"
+            )
+
+        foreground_prob_map = foreground_prob_map >= self.threshold  # uncertain area
+
+        marker = mask - convert_to_dst_type(foreground_prob_map, mask, np.uint8)[0]  # certain foreground
```

Comment:
@KumoLiu, I very much like that you have have added comments about uncertain and certain area but it seems that they author of this model have made some heuristic choices that might not match exactly with the concept of foreground probability or uncertainty. 

This implementation [here](https://github.com/Dana-Farber-AIOS/pathml/blob/d5982f4b84708871e635c3db403dec3887021343/pathml/ml/hovernet.py#L725-L731) makes more sense to me than the original hovernet implementation. What do you think?

Reply:
Hi @drbeh, it seems the implementation [you shared](https://github.com/Dana-Farber-AIOS/pathml/blob/d5982f4b84708871e635c3db403dec3887021343/pathml/ml/hovernet.py#L725-L731) has a little difference complementation with the original repo, do you mean you want me to implement in that way? `forground_prob_map` indeed lead some misunderstanding, I would suggest to use `GetAdjacentArea` to replace `GenerateProbabilityMap` and `adjacent_area` to replace `foreground_prob_map` because I think it indeed finds the neighboring area between nucleus. 
 Here is what the original paper defines this concept. "Sm highlights areas where there is a significant difference in neighbouring pixels within the horizontal and vertical maps"
Or maybe you have other suggestions, thanks!